src/hint_processor/builtin_hint_processor/field_arithmetic.rs

```diff
@@ -0,0 +1,271 @@
+};
+
+use super::hint_utils::{get_relocatable_from_var_name, insert_value_from_var_name};
+use super::secp::bigint_utils::BigInt3;
+use super::uint384::{pack, split};
+/* Implements Hint:
+      %{
+           from starkware.python.math_utils import is_quad_residue, sqrt
+
+           def split(num: int, num_bits_shift: int = 128, length: int = 3):
+               a = []
+               for _ in range(length):
+                   a.append( num & ((1 << num_bits_shift) - 1) )
+                   num = num >> num_bits_shift
+               return tuple(a)
+
+           def pack(z, num_bits_shift: int = 128) -> int:
+               limbs = (z.d0, z.d1, z.d2)
+               return sum(limb << (num_bits_shift * i) for i, limb in enumerate(limbs))
+
+
+           generator = pack(ids.generator)
+           x = pack(ids.x)
+           p = pack(ids.p)
+
+           success_x = is_quad_residue(x, p)
+           root_x = sqrt(x, p) if success_x else None
+
+           success_gx = is_quad_residue(generator*x, p)
+           root_gx = sqrt(generator*x, p) if success_gx else None
+
+           # Check that one is 0 and the other is 1
+           if x != 0:
+               assert success_x + success_gx ==1
+
+           # `None` means that no root was found, but we need to transform these into a felt no matter what
+           if root_x == None:
+               root_x = 0
+           if root_gx == None:
+               root_gx = 0
+           ids.success_x = int(success_x)
+           split_root_x = split(root_x)
+           split_root_gx = split(root_gx)
+           ids.sqrt_x.d0 = split_root_x[0]
+           ids.sqrt_x.d1 = split_root_x[1]
+           ids.sqrt_x.d2 = split_root_x[2]
+           ids.sqrt_gx.d0 = split_root_gx[0]
+           ids.sqrt_gx.d1 = split_root_gx[1]
+           ids.sqrt_gx.d2 = split_root_gx[2]
+       %}
+*/
+pub fn get_square_root(
+    vm: &mut VirtualMachine,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+) -> Result<(), HintError> {
+    let sqrt_x_addr = get_relocatable_from_var_name("sqrt_x", vm, ids_data, ap_tracking)?;
+    let sqrt_gx_addr = get_relocatable_from_var_name("sqrt_gx", vm, ids_data, ap_tracking)?;
+    let generator = pack(
+        BigInt3::from_var_name("generator", vm, ids_data, ap_tracking)?,
+        128,
+    );
+    let x = pack(BigInt3::from_var_name("x", vm, ids_data, ap_tracking)?, 128);
+    let p = pack(BigInt3::from_var_name("p", vm, ids_data, ap_tracking)?, 128);
+    let success_x = is_quad_residue(&x, &p)?;
+
+    let root_x = if success_x {
+        sqrt_prime_power(&x, &p).unwrap_or_default()
+    } else {
+        BigUint::zero()
+    };
+
+    let gx = generator * &x;
+    let success_gx = is_quad_residue(&gx, &p)?;
+
+    let root_gx = if success_gx {
+        sqrt_prime_power(&gx, &p).unwrap_or_default()
+    } else {
+        BigUint::zero()
+    };
+
+    if !&x.is_zero() && !(success_x as u8 + success_gx as u8).is_one() {
```

Comment:
I'm not sure this will behave exactly as the Python code. At this point, in the original code `x` could be `None`, which doesn't compare as equal to `0`. In the hint `x` is converted to `0` when originally `None` after the check.

Reply:
It would help if you could add an integration test for this to ensure that case fails or succeeds the same way (that is, you need `x` that isn't a quad residue (so it evaluates to `None` in Python) and `gx` that isn't either. In Python, I think, the check will fail (`x != 0` means we evaluate the `assert`, and because none of the values were quad residues the sum will be `0` and the `assert` fails), while in our code this `None` became `0`, so `!&x.is_zero()` evaluate to `false` and we don't reach the `assert`.

Reply:
Ignore this, `root_x` is what could be `None`, not `x`.